docs/advanced/optimizing-selectors.md

```diff
@@ -0,0 +1,110 @@
+
+One way to solve this problem is to turn off the `injectContainerState` selector option. This enables you to explicitly specify arguments using the `@Selector` decorator. For example:
+
+```ts
+@Selector([SomeState])
+static getData(state: SomeStateModel) {
+   return state.data;
+}
+
+@Selector([SomeState.getData])
+static getViewData(data: Data[]) {
+   return data.map(d => expensiveFunction(d));
+}
+```
+
+This `getViewData` selector will not be recalculated when a user types into the input component. This selector targets the specific property of `state` it cares about as its argument by leveraging an additional selector. When the `name` property of state changes, the `getViewData` arguments _do not change_. Memoization is taken advantage of.
+
+An alternative solution to turning off the selector option is to create selectors outside of your state classes.
+
+```ts
+@State({...})
+@Injectable()
+export class SomeState {
+  @Selector()
+  static getData(state: SomeStateModel) {
+    return state.data;
+  }
+}
+
+@Injectable()
+export class SomeSelectorService {
+  @Selector([SomeState.getData])
+  static getViewData(data: Data[]) {
+    return data.map(d => expensiveFunction(d));
+  }
+}
+```
+
+## Implementation
+
+Selectors are calculated when state changes. As your application grows, the number of state changes increases. Finding optimizations in your selector implementations can have significant benefits.
+
+For example, say you have this state model:
+
+```ts
+interface SelectedDataStateModel {
+  selectedIds: number[];
+}
+```
+
+And you have this selector:
+
+```ts
+@Selector([SelectedDataState])
+isDataSelected(state: SelectedDataStateModel) {
+  return (id: number) => state.selectedIds.includes(id);
+}
+```
+
+The selector returns a function, which accepts an id as an argument and returns a boolean indicating whether or not this id is selected. In this example, we want to render a list of checkboxes.
+
+```html
+<ng-container *ngIf="isDataSelected$ | async as isDataSelected">
+  <data-check-box
+    *ngFor="data of data$ | async"
+    [checked]="isDataSelected(d.id)"
+  ></data-check-box>
+</ng-container>
+```
+
+When users check or uncheck an item, `state.selectedIds` is updated, therefore the `isDataSelected` selector is recalculated and the list must re-render. Every time the list re-renders, `isDataSelected` is invoked `data.length` number of times. Because the selector implementation uses [Array.includes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes). This template renders in `O(n^2)` time complexity - **Ugh!**. One magnitude of `n` for the length of data , another for `state.selectedIds.length`.
+
+Here's one way to improve performance in that example:
+
+```ts
+@Selector([SelectedDataState])
+isDataSelected(state: SelectedDataStateModel) {
+  const selectedIds = new Set(state.selectedIds);
+  return (id: number) => selectedIds.has(id);
```

Comment:
I think a comment here would be useful to explicitly point out the fact that you are returning a closure over the `selectedIds` variable. It is not immediately obvious why this is better.

Reply:
You’re absolutely correct.  I will make note of this.  Thanks for the feedback!